File: intellij-plugin/fixtures/com/ezevillo/gitreview/fixtures/PanelFixtures.kt
package com.ezevillo.gitreview.fixtures

import com.ezevillo.gitreview.domain.BranchFinish
import com.ezevillo.gitreview.domain.BranchRecord
import com.ezevillo.gitreview.domain.EffectiveConfig
import com.ezevillo.gitreview.domain.EntryRecord
import com.ezevillo.gitreview.domain.PanelInputs
import com.ezevillo.gitreview.domain.PanelModel
import com.ezevillo.gitreview.domain.PanelWhy
import com.ezevillo.gitreview.domain.PendingFinish
import com.ezevillo.gitreview.domain.ReviewMode
import com.ezevillo.gitreview.domain.ReviewState
import com.ezevillo.gitreview.domain.Situation
import com.ezevillo.gitreview.domain.StateRecord
import com.ezevillo.gitreview.domain.WalkthroughStatus
import com.ezevillo.gitreview.domain.WhyState
import com.ezevillo.gitreview.domain.buildPanelModel
import com.ezevillo.gitreview.domain.parseListPorcelain
import com.ezevillo.gitreview.domain.parsePorcelain
import com.ezevillo.gitreview.domain.toPathRef

/**
 * Shared [PanelModel] fixtures for tests and `runPanelPreview`.
 * Built from porcelain sample text via [parsePorcelain] / [buildPanelModel].
 */
object PanelFixtures {
    fun all(): List<Pair<String, PanelModel>> = listOf(
        "cli-missing" to cliMissing(),
        "cli-outdated" to cliOutdated(),
        "no-review setup" to noReviewSetup(),
        "no-review ready" to noReviewReady(),
        "finish-pending" to finishPending(),
        "out-of-range" to outOfRange(),
        "error" to error(),
        "review walk" to reviewWalk(),
        "review step" to reviewStep(),
        "review whole" to reviewWhole(),
        "finish-conflict" to finishConflict(),
        "review walk draft" to reviewWalkDraft(),
        "review walk busy" to reviewWalk(busy = true),
        "review walk empty cursor" to reviewWalkEmptyCursor(),
        "review whole empty" to reviewWholeEmpty(),
    )

    fun cliMissing(): PanelModel = buildPanelModel(
        ReviewState(situation = Situation.CLI_MISSING, stderr = "not found"),
        PanelInputs(busy = false),
    )

    fun cliOutdated(): PanelModel = buildPanelModel(
        ReviewState(situation = Situation.CLI_OUTDATED, stderr = "0.3.0"),
        PanelInputs(busy = false),
    )

    fun noReviewSetup(): PanelModel = buildPanelModel(
        ReviewState(
            situation = Situation.NO_REVIEW,
            config = EffectiveConfig(base = null, remote = "origin"),
        ),
        PanelInputs(busy = false),
    )

    fun noReviewReady(): PanelModel {
        val listPorcelain = """
            branch	review-saved/feature	1	0	0	walk	2	5
            branch	review/other	0	0	1	step
        """.trimIndent()
        return buildPanelModel(
            ReviewState(
                situation = Situation.NO_REVIEW,
                config = EffectiveConfig(base = "main", remote = "origin"),
                branches = parseListPorcelain(listPorcelain),
            ),
            PanelInputs(busy = false),
        )
    }

    fun finishPending(): PanelModel {
        val branches = listOf(
            BranchRecord(
                name = "review/feature",
                saved = false,
                current = true,
                orphan = false,
                finish = BranchFinish(state = "pending", onto = false),
            ),
        )
        return buildPanelModel(
            ReviewState(situation = Situation.FINISH_PENDING, branches = branches),
            PanelInputs(busy = false),
        ).let { m ->
            if (m.pendingFinish != null) m
            else m.copy(pendingFinish = PendingFinish(branch = "review/feature", onto = false))
        }
    }

    fun outOfRange(): PanelModel = buildPanelModel(
        ReviewState(situation = Situation.OUT_OF_RANGE, stderr = "base moved"),
        PanelInputs(busy = false),
    )

    fun error(): PanelModel = buildPanelModel(
        ReviewState(
            situation = Situation.ERROR,
            stderr = "Open a single-folder workspace that is a git repository. git review uses one root (like the CLI cwd); multi-root is not supported.",
        ),
        PanelInputs(busy = false),
    )

    fun reviewWalk(
        busy: Boolean = false,
        why: PanelWhy = PanelWhy(WhyState.PRESENT, "Because it matters."),
        atFirst: Boolean = true,
        atLast: Boolean = false,
        position: Int = 1,
    ): PanelModel {
        val walkPorcelain = """
            state	review/feature	feature	deadbeefcafebabe	walk	applied	$position	3	3	"src/a.kt"	1
            entry	1	src/a.kt	1	1
            entry	2	src/b.kt	0	1
            entry	3	src/c.kt	0	0
        """.trimIndent()
        val walkParsed = parsePorcelain(walkPorcelain)
        val walkState = ReviewState(
            situation = Situation.REVIEW,
            state = walkParsed.state,
            entries = walkParsed.entries,
        )
        val model = buildPanelModel(walkState, PanelInputs(busy = busy, why = why))
        return model.copy(atFirst = atFirst, atLast = atLast || position >= (model.total ?: 0))
    }

    /**
     * 011: la misma review walk, leída sobre el borrador del revisor. El
     * registro `draft` viaja por el porcelain como cualquier otro, así que el
     * badge sale del mismo camino que en la extensión.
     */
    fun reviewWalkDraft(): PanelModel {
        val porcelain = """
            state	review/feature	feature	deadbeefcafebabe	walk	applied	1	3	3	"src/a.kt"	0
            entry	1	src/a.kt	0	1
            entry	2	src/b.kt	0	1
            entry	3	src/c.kt	0	0
            draft
        """.trimIndent()
        val parsed = parsePorcelain(porcelain)
        val state = ReviewState(
            situation = Situation.REVIEW,
            state = parsed.state,
            entries = parsed.entries,
            draft = parsed.draft,
        )
        val model = buildPanelModel(
            state,
            PanelInputs(busy = false, why = PanelWhy(WhyState.PRESENT, "Because I read it first.")),
        )
        return model.copy(atFirst = true, atLast = false)
    }

    fun reviewWalkEmptyCursor(): PanelModel {
        val m = reviewWalk(position = 1)
        return m.copy(current = null, position = 99, total = 3)
    }

    fun reviewStep(
        busy: Boolean = false,
        position: Int = 2,
        withFiles: Boolean = true,
    ): PanelModel {
        val stepFiles = if (withFiles) {
            listOf(
                EntryRecord(1, toPathRef("src/a.kt")),
                EntryRecord(2, toPathRef("src/b.kt")),
            )
        } else {
            emptyList()
        }
        return buildPanelModel(
            ReviewState(
                situation = Situation.REVIEW,
                state = StateRecord(
                    "review/f", "f", "tipsha01", ReviewMode.STEP, WalkthroughStatus.NONE,
                    position = position, total = 4, recorded = 4, current = "abc1234",
                ),
                entries = listOf(
                    EntryRecord(1, "aaa1111", banked = false),
                    EntryRecord(2, "abc1234", banked = true),
                    EntryRecord(3, "ccc3333", banked = false),
                    EntryRecord(4, "ddd4444", banked = false),
                ),
                files = stepFiles,
                subjects = mapOf(2 to "Fix the thing"),
                authors = mapOf(2 to "Ada"),
            ),
            PanelInputs(
                busy = busy,
                lastOpened = if (withFiles) "src/a.kt" else null,
            ),
        )
    }

    fun reviewWhole(fileCount: Int = 2): PanelModel {
        val entries = (1..fileCount).map { i ->
            EntryRecord(i, toPathRef("file$i.kt"))
        }
        return buildPanelModel(
            ReviewState(
                situation = Situation.REVIEW,
                state = StateRecord(
                    "review/f", "f", "tipsha01", ReviewMode.WHOLE, WalkthroughStatus.NONE,
                ),
                entries = entries,
                base = "main",
            ),
            PanelInputs(
                busy = false,
                lastOpened = entries.firstOrNull()?.let {
                    when (val id = it.id) {
                        is com.ezevillo.gitreview.domain.PathRef -> id.display
                        else -> id.toString()
                    }
                },
            ),
        )
    }

    fun reviewWholeEmpty(): PanelModel = buildPanelModel(
        ReviewState(
            situation = Situation.REVIEW,
            state = StateRecord(
                "review/f", "f", "tipsha01", ReviewMode.WHOLE, WalkthroughStatus.NONE,
            ),
            entries = emptyList(),
            base = "main",
        ),
        PanelInputs(busy = false),
    )

    fun finishConflict(): PanelModel = buildPanelModel(
        ReviewState(
            situation = Situation.FINISH_CONFLICT,
            state = StateRecord(
                "review/f", "f", "tipsha01", ReviewMode.WALK, WalkthroughStatus.APPLIED,
                position = 1, total = 2, recorded = 2, current = toPathRef("a.kt"),
            ),
            entries = listOf(EntryRecord(1, toPathRef("a.kt"), essential = true, annotated = true)),
        ),
        PanelInputs(busy = false, why = PanelWhy(WhyState.PRESENT, "conflict entry")),
    )
}

File: visualstudio-extension/fixtures/PanelFixtures.cs
using GitReview.Domain;

namespace GitReview.Fixtures;

/// <summary>
/// Panel models for every situation, built from porcelain the way the host builds
/// them. Shared by the xUnit suite and by <c>--preview</c> / <c>--verify</c>, the
/// same way <c>jetbrains-plugin/fixtures/</c> is shared by that client's tests and
/// its <c>runPanelPreview</c>: two sets would let the gallery and the gate drift,
/// and the situations only the gallery had were the ones no test ever asserted on.
/// Compiled into both projects by <c>Compile Include</c>; it is domain-only code,
/// with no reference to WPF or the VSSDK.
/// </summary>
public static class PanelFixtures
{
    public static IReadOnlyList<(string Name, PanelModel Model)> All() => new[]
    {
        ("cli-missing", CliMissing()),
        ("cli-outdated", CliOutdated()),
        ("no-review setup", NoReviewSetup()),
        ("no-review ready", NoReviewReady()),
        ("no-review drafts", NoReviewDrafts()),
        ("no-review empty", NoReviewEmpty()),
        ("finish-pending", FinishPending()),
        ("out-of-range", OutOfRange()),
        ("error", Error()),
        ("review walk", ReviewWalk()),
        ("review step", ReviewStep()),
        ("review whole", ReviewWhole()),
        ("finish-conflict", FinishConflict()),
        ("review walk draft", ReviewWalkDraft()),
        ("review walk busy", ReviewWalk(busy: true)),
        ("review walk empty cursor", ReviewWalkEmptyCursor()),
        ("review whole empty", ReviewWholeEmpty()),
    };

    public static PanelModel CliMissing() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(Situation.CliMissing, Stderr: "not found"),
        new PanelInputs(false));

    public static PanelModel CliOutdated() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(Situation.CliOutdated, Stderr: "0.3.0"),
        new PanelInputs(false));

    public static PanelModel NoReviewSetup() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(Situation.NoReview, Config: new EffectiveConfig(null, "origin")),
        new PanelInputs(false));

    public static PanelModel NoReviewReady()
    {
        var listPorcelain = "branch\treview-saved/feature\t1\t0\t0\twalk\t2\t5\nbranch\treview/other\t0\t0\t1\tstep";
        return PanelModelBuilder.BuildPanelModel(
            new ReviewState(
                Situation.NoReview,
                Config: new EffectiveConfig("main", "origin"),
                Branches: Porcelain.ParseListPorcelain(listPorcelain)),
            new PanelInputs(false));
    }

    /// <summary>
    /// Two reading orders started and not paused, plus the inventory below. The
    /// second row does NOT offer "Validate and start": its instruction block was
    /// deleted by hand, so the CLI reports `unknown` and the flags cannot be
    /// replicated.
    /// </summary>
    public static PanelModel NoReviewDrafts(bool busy = false)
    {
        var cfg =
            "draft\tfeature/telemetry\t/repo/.git/review-walkthrough/feature/telemetry.md\t3\t9\tlocal\tdelta\n" +
            "draft\tfeature/pagos\t/repo/.git/review-walkthrough/feature/pagos.md\t0\t5\tunknown\tunknown\n" +
            "draft\tfeature/legacy\t/repo/.git/review-walkthrough/feature/legacy.md\t1\t1\tremote\tfull\n";
        return PanelModelBuilder.BuildPanelModel(
            new ReviewState(
                Situation.NoReview,
                Config: new EffectiveConfig("main", "origin"),
                Branches: Porcelain.ParseListPorcelain("branch\treview-saved/feature\t1\t0\t0\twalk\t2\t5"),
                Drafts: ConfigPorcelain.ParseConfigPorcelain(cfg).Drafts),
            new PanelInputs(busy));
    }

    /// <summary>The same state with a mutation in flight: the one thing that switches the row off.</summary>
    public static PanelModel NoReviewDraftsBusy() => NoReviewDrafts(busy: true);

    /// <summary>Configured, but nothing left over from an earlier review.</summary>
    public static PanelModel NoReviewEmpty() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(
            Situation.NoReview,
            Config: new EffectiveConfig("main", "origin"),
            Branches: Array.Empty<BranchRecord>()),
        new PanelInputs(false));

    public static PanelModel FinishPending()
    {
        var branches = new[]
        {
            new BranchRecord("review/feature", false, true, false,
                Finish: new BranchFinish("pending", false)),
        };
        return PanelModelBuilder.BuildPanelModel(
            new ReviewState(Situation.FinishPending, Branches: branches),
            new PanelInputs(false));
    }

    public static PanelModel OutOfRange() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(Situation.OutOfRange, Stderr: "base moved"),
        new PanelInputs(false));

    public static PanelModel Error() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(
            Situation.Error,
            Stderr: "Open a single-folder workspace that is a git repository. git review uses one root (like the CLI cwd); multi-root is not supported."),
        new PanelInputs(false));

    public static PanelModel ReviewWalk(
        bool busy = false,
        PanelWhy? why = null,
        bool atFirst = true,
        bool atLast = false,
        int position = 1)
    {
        why ??= new PanelWhy(WhyState.Present, "Because it matters.");
        var walkPorcelain =
            $"state\treview/feature\tfeature\tdeadbeefcafebabe\twalk\tapplied\t{position}\t3\t3\t\"src/a.kt\"\t1\n" +
            "entry\t1\tsrc/a.kt\t1\t1\nentry\t2\tsrc/b.kt\t0\t1\nentry\t3\tsrc/c.kt\t0\t0";
        var walkParsed = Porcelain.ParsePorcelain(walkPorcelain);
        var walkState = new ReviewState(
            Situation.Review,
            State: walkParsed.State,
            Entries: walkParsed.Entries);
        var model = PanelModelBuilder.BuildPanelModel(walkState, new PanelInputs(busy, Why: why));
        return model with { AtFirst = atFirst, AtLast = atLast || position >= (model.Total ?? 0) };
    }

    public static PanelModel ReviewWalkDraft()
    {
        var porcelain =
            "state\treview/feature\tfeature\tdeadbeefcafebabe\twalk\tapplied\t1\t3\t3\t\"src/a.kt\"\t0\n" +
            "entry\t1\tsrc/a.kt\t0\t1\nentry\t2\tsrc/b.kt\t0\t1\nentry\t3\tsrc/c.kt\t0\t0\ndraft";
        var parsed = Porcelain.ParsePorcelain(porcelain);
        var state = new ReviewState(
            Situation.Review,
            State: parsed.State,
            Entries: parsed.Entries,
            Draft: parsed.Draft);
        var model = PanelModelBuilder.BuildPanelModel(
            state,
            new PanelInputs(false, Why: new PanelWhy(WhyState.Present, "Because I read it first.")));
        return model with { AtFirst = true, AtLast = false };
    }

    /// <summary>
    /// The cursor points at a position no entry has — what a walkthrough that no
    /// longer intersects the range leaves behind.
    /// </summary>
    public static PanelModel ReviewWalkEmptyCursor() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(
            Situation.Review,
            State: new StateRecord(
                "review/feature", "feature", "deadbeef", ReviewMode.Walk, WalkthroughStatus.Applied,
                Position: 9, Total: 3, Recorded: 3, Current: Unquote.ToPathRef("src/gone.kt")),
            Entries: new[]
            {
                new EntryRecord(1, Unquote.ToPathRef("src/a.kt"), Essential: false, Annotated: true),
                new EntryRecord(2, Unquote.ToPathRef("src/b.kt"), Essential: false, Annotated: true),
                new EntryRecord(3, Unquote.ToPathRef("src/c.kt"), Essential: false, Annotated: true),
            }),
        new PanelInputs(false));

    public static PanelModel ReviewStep(bool busy = false, int position = 2, bool withFiles = true)
    {
        var stepFiles = withFiles
            ? new[]
            {
                new EntryRecord(1, Unquote.ToPathRef("src/a.kt")),
                new EntryRecord(2, Unquote.ToPathRef("src/b.kt")),
            }
            : Array.Empty<EntryRecord>();
        return PanelModelBuilder.BuildPanelModel(
            new ReviewState(
                Situation.Review,
                State: new StateRecord(
                    "review/f", "f", "tipsha01", ReviewMode.Step, WalkthroughStatus.None,
                    Position: position, Total: 4, Recorded: 4, Current: "abc1234"),
                Entries: new[]
                {
                    new EntryRecord(1, "aaa1111", Banked: false),
                    new EntryRecord(2, "abc1234", Banked: true),
                    new EntryRecord(3, "ccc3333", Banked: false),
                    new EntryRecord(4, "ddd4444", Banked: false),
                },
                Files: stepFiles,
                Subjects: new Dictionary<int, string> { [2] = "Fix the thing" },
                Authors: new Dictionary<int, string> { [2] = "Ada" }),
            new PanelInputs(busy, LastOpened: withFiles ? "src/a.kt" : null));
    }

    public static PanelModel ReviewWhole(int fileCount = 2)
    {
        var entries = Enumerable.Range(1, fileCount)
            .Select(i => new EntryRecord(i, Unquote.ToPathRef($"file{i}.kt")))
            .ToList();
        return PanelModelBuilder.BuildPanelModel(
            new ReviewState(
                Situation.Review,
                State: new StateRecord("review/f", "f", "tipsha01", ReviewMode.Whole, WalkthroughStatus.None),
                Entries: entries,
                Base: "main"),
            new PanelInputs(false, LastOpened: entries.Count > 0
                ? ((PathRef)entries[0].Id).Display
                : null));
    }

    public static PanelModel ReviewWholeEmpty() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(
            Situation.Review,
            State: new StateRecord("review/f", "f", "tipsha01", ReviewMode.Whole, WalkthroughStatus.None),
            Entries: Array.Empty<EntryRecord>(),
            Base: "main"),
        new PanelInputs(false));

    public static PanelModel FinishConflict() => PanelModelBuilder.BuildPanelModel(
        new ReviewState(
            Situation.FinishConflict,
            State: new StateRecord(
                "review/feature", "feature", "deadbeef", ReviewMode.Walk, WalkthroughStatus.Applied,
                Position: 1, Total: 2, Recorded: 2, Current: Unquote.ToPathRef("src/a.kt")),
            Entries: new[]
            {
                new EntryRecord(1, Unquote.ToPathRef("src/a.kt"), Essential: true, Annotated: true),
                new EntryRecord(2, Unquote.ToPathRef("src/b.kt"), Essential: false, Annotated: true),
            },
            Finish: new StatusFinishRecord("conflict", false)),
        new PanelInputs(false, Why: new PanelWhy(WhyState.Present, "Why text.")));
}
